### PR TITLE
ENH: Vendor meson for multi-target build support

### PR DESCRIPTION
**Pull request by rgommers**
This is a much more minimal alternative to gh-24365. It only vendors Meson, and uses PATH manipulation to get `meson-python` and `spin` to pick up our vendored Meson - which is a friendly fork which lives at https://github.com/numpy/meson and is meant to temporarily include the feature we need for SIMD support.

Tested in combination with gh-23096, and it looks like everything works. Let's see how happy CI is ....

### CODE REVIEW

**Comment by rgommers** [comment]
Already one weird failure, on the `musllinux` job:
```
picking up the wrong `meson`: ['/__w/numpy/numpy/test_env/lib/python3.10/site-packages/mesonbuild']
```
that `__w` path looks very odd, doesn't immediately ring a bell.

File: vendored-meson/entrypoint/meson [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import re
+import sys
+import pathlib
+
+
+# The numpy-vendored version of Meson
+meson_dir = str(pathlib.Path(__file__).resolve().parent.parent / 'meson')
+sys.path.insert(0, meson_dir)
```
Comment (mattip) [inline vendored-meson/entrypoint/meson:10]: Maybe assert that this directory exists?

**Comment by mattip** [comment]
I think the `_w` is a false flag, it has to do with the container file system. It seems more like there is a problem with `sys.path` injection, perhaps the `git submodule update --init` did not work.

**Comment by mattip** [comment]
There is an [error in the log](https://github.com/numpy/numpy/actions/runs/5813406685/job/15760891310?pr=24379#step:3:34)
```
Could not access submodule 'vendored-meson/meson' at commit 45e98f381
```

**Comment by rgommers** [comment]
> `Could not access submodule 'vendored-meson/meson' at commit 45e98f381`

Good catch - that is a little curious. Guess I'll need to add a check for that. I fixed the other (lint, sdist) failures. Let's see if it happens again. The other submodules did get initialized, so it's unclear to me what the problem was there.

**Comment by mattip** [comment]
musl linux is cloning, and then fetching this PR (which has a new submodule), but where is the final `git submodule update --init` after the checkout? https://github.com/numpy/numpy/blob/60c8b6a9f467f7809d4078aec6e4fe7cbc8191b8/.github/workflows/linux_musl.yml#L36-L46

File: tools/travis-test.sh [hunk 2]
```diff
@@ -223,7 +223,7 @@ elif [ -n "$USE_SDIST" ] && [ $# -eq 0 ]; then
   $PYTHON -c "import fcntl; fcntl.fcntl(1, fcntl.F_SETFL, 0)"
   # ensure some warnings are not issued
   export CFLAGS=$CFLAGS" -Wno-sign-compare -Wno-unused-result -Wno-error=undef"
-  $PYTHON setup.py sdist
+  $PYTHON -m build --sdist
```
Comment (mattip) [inline tools/travis-test.sh:226]: Is `build` installed? The [CI fails](https://github.com/numpy/numpy/actions/runs/5813680535/job/15761994512?pr=24379#step:4:333) with
```
+ python -m build --sdist
/home/runner/work/numpy/numpy/builds/venv/bin/python: No module named build
```

**Comment by rgommers** [comment]
> musl linux is cloning, and then fetching this PR (which has a new submodule), but where is the final `git submodule update --init` after the checkout?

We didn't need it until this PR. I'd expect this to break downstream CI in places as well, but nothing we can do about that beyond a clear error message.

Also, sigh:
```python
        # using git commands to clone because versioneer doesn't work when
        # actions/checkout is used for the clone step in a container
```
how things go in circles ...

File: vendored-meson/entrypoint/meson (hunk 1 above)
Comment (rgommers) [reply vendored-meson/entrypoint/meson:10]: done now

**Comment by mattip** [comment]
> We didn't need it until this PR.

I think the problem here is we are adding a new submodule in the PR. Perhaps the previous times we added a submodule (for svml, x86-simd-sort) we didn't check/care that fetching the submodule failed.

**Comment by rgommers** [comment]
> I think the problem here is we are adding a new submodule in the PR. Perhaps the previous times we added a submodule (for svml, x86-simd-sort) we didn't check/care that fetching the submodule failed.

Indeed - that's part of the tradeoff of git submodule vs. inclusion of sources. I think we should simply accept it though, it's part of the deal.

This looks ready now, all CI seems happy (the one TravisCI failure is the regular ppc64le job abort). @mattip if this looks good to you, how about we merge this and then I'll try and get the PR to enable SIMD support merge today too? That plus final tweaks to versioneer removal should get us to where we can do a first 1.26.0 pre-release.

**Comment by mattip** [comment]
Thanks @rgommers

**Comment by rgommers** [comment]
@charris I suggest to wait with backporting anything until we've got gh-23096 in too. It's possible a few tweaks are still required, we'll see.

**Comment by rgommers** [comment]
I spent ~ half a day debugging on CI and investigating the problem on Windows. It seems impossible to do this via PATH manipulation. The `meson` script with the executable bit set is not running, even when you put it right next to the `__init__.py` of the in-tree build backend or use the full path to it. On the other hand `python /path/to/entrypoint/meson` runs perfectly fine. This is very simple to do with a fork of `meson-python`, which I had initially.

Unless someone (@eli-schwartz or @mattip maybe?) has an answer to this, I'll go back to a modified `meson-python` and will include it as a git submodule too.

**Comment by rgommers** [comment]
> I guess you could add upstream meson-python support for setting `$MESON` similarly to `$NINJA`, then very quickly release a new meson-python point release.

Yes, I think I'd like to add support for `$MESON`, which is a good idea in general. But at a gentler pace; if we get that before the final 1.26.0 release then that should be fine.

**Comment by mattip** [comment]
> Some debug printing in CI 

Which CI run?

**Comment by rgommers** [comment]
> Which CI run?

There's ~10 failed builds with different invocations at https://github.com/rgommers/numpy/actions. I'd advise not digging into those, because they won't tell you much beyond "yes the paths were right, and no the `meson` executable never runs".

My current plan is:
1. Fork `meson-python` and add it as a git submodule in a similar way as done for `meson` in this PR
2. Get gh-23096 merged
3. Backport to 1.26.x for a pre-release
4. Add `$MESON` support for `meson-python` as Eli suggested
5. Do a `meson-python` release, then undo the `meson-python` vendoring

**Comment by mattip** [comment]
Assuming you are in a venv (so sys.executable is in scripts), maybe copy the meson.exe from `python -c "import os, sys; print(os.path.dirname(sys.executable))` to `vendored-meson\entrypoint` ?

**Comment by rgommers** [comment]
I've tried multiple variations like that already. And in venv or not doesn't matter (and we can't assume anything about that anyway, it should work robustly no matter the type of env).

**Comment by eli-schwartz** [comment]
It's very much not worth the fuss. Do note however that the meson.exe files are different -- that PE executable is created by taking the base launcher executable vendored in distlib, and appending a zipapp with a `path/to/python.exe` shebang and a `__main__.py` which is a renamed script entrypoint (usually does the `importlib.metadata.distribution(XXX).entry_points` dance). It doesn't contain the special sys.path munging that meson's git repository wrapper script has.

**Comment by mattip** [comment]
one option would be to `cd` into vendored-meson/meson` and do `pip install .`

**Comment by rgommers** [comment]
That's a no no, we can't just go install a `meson` package into the active env that the user did not ask for. We cannot make any permanent changes to the build env.

**Comment by rgommers** [comment]
gh-24365 was actually the correct and pragmatic approach, modulo that I should have used git submodules there.

**Comment by charris** [comment]
> pragmatic approach

 It doesn't need to be beautiful as long as it gets the job done :)

**Comment by rgommers** [comment]
> Do note however that the meson.exe files are different

That's a point of complexity for the `$MESON` idea probably, right? I wouldn't want to deal with that, but rather have a rule that of `$MESON` contains a name ending in `.py`, it'll be run by meson-python with `sys.executable` prepended - that should be much easier and more robust.

**Comment by eli-schwartz** [comment]
Right, a .py script is the logical way to do this. It should be safely assumed that if the entrypoint contains an explicit `.py` then it is not a $PATH installed tool expecting to import `mesonbuild.mesonmain:main()` from site-packages, and rather it's a setup script that sets its own path up. This also makes sure that things work on e.g. FreeBSD, where `#!/usr/bin/env python3` does not work because some clever fellows decided that it is "unnecessary" for software installed via the BSD ports tree and containing `/usr/local/bin/python3.9` shebangs or whatever version of python they have in the /usr/local hierarchy at any given time.

(Apparently there's an optional package containing the "python3" symlink for people that really really want it, but you need to know you need it.)

The other possible approach is documenting that the `$MESON` variable is handled like `$CC` -- as a space-separated list of arguments e.g. `python.exe vendored-meson\\entrypoint\\meson.py`. But I don't think there's any real advantage there.

**Comment by rgommers** [comment]
Thanks. Agreed, both of those should do the job. I'll open the feature request soon and will try to get this into the next version.

> (Apparently there's an optional package containing the "python3" symlink for people that really really want it, but you need to know you need it

🤦🏼

**Pull request merged**